File: data-structures/array/scratchpad.rb
def exists?(array, item)
  array.each do |value|
    return true if value == item
  end

  false
end

def linear_search(input_array, search_value)
  input_array.each_with_index do |value, index|
    return index if value == search_value
  end

  nil
end

def binary_search(input_array, search_value)
  lower_bound = 0
  upper_bound = input_array.size - 1

  while lower_bound <= upper_bound
    midpoint = (lower_bound + upper_bound) / 2
    element = input_array[midpoint]

    if search_value == element
      return midpoint
    end

    if element < search_value
      lower_bound = midpoint + 1
    end

    if element > search_value
      upper_bound = midpoint - 1
    end 
  end

  nil
end

File: data-structures/array/scratchpad.go
package main

import (
	"errors"
	"fmt"
)

func exists(inputArray []string, item string) bool {
	for _, value := range inputArray {
		if value == item {
			return true
		}
	}
	return false
}

func linearSearch(inputArray []int, searchValue int) (int, error) {
	for index, value := range inputArray {
		if value == searchValue {
			return index, nil
		}
	}
	return 0, errors.New("Not Exists")
}

func binarySearch(inputArray []int, searchValue int) (int, error) {
	lowerBound := 0
	upperBound := len(inputArray) - 1

	for lowerBound <= upperBound {
		midpoint := (lowerBound + upperBound) / 2
		element := inputArray[midpoint]

		if searchValue == element {
			return midpoint, nil
		}

		if element < searchValue {
			lowerBound = midpoint + 1
		}

		if element > searchValue {
			upperBound = midpoint - 1
		}
	}

	return 0, errors.New("Not Exists")
}

func main() {
	cars := []string{"Mazda", "BWM", "Rolls Royce"}
	fmt.Println(exists(cars, "Bentley"))
	fmt.Println(exists(cars, "Rolls Royce"))

	var result int
	var err error
	numbers := []int{3, 7, 15, 23, 46, 59, 63, 72, 88, 95}
	result, err = linearSearch(numbers, 100)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(result)
	}

	result, err = linearSearch(numbers, 72)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(result)
	}

	result, err = binarySearch(numbers, 100)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(result)
	}

	result, err = binarySearch(numbers, 72)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(result)
	}
}
